R/srcrefs.R

```diff
@@ -75,33 +79,57 @@ srcrefs.default <- function(x, ..., srcref_names = NULL) {
 
 #' @exportS3Method
 #' @rdname srcrefs
-srcrefs.list <- function(x, ..., srcref_names = NULL) {
-  flat_map_srcrefs(x)
+srcrefs.list <- function(x, srcref_names = NULL, breadcrumbs = character()) {
+  flat_map_srcrefs(x, ns = srcref_names, breadcrumbs = breadcrumbs)
 }
 
 #' @exportS3Method
 #' @rdname srcrefs
-srcrefs.namespace <- function(x, ...) {
-  flat_map_srcrefs(as.list(x, all.names = TRUE), ns = env_ns_name(x))
+srcrefs.namespace <- function(x, ..., breadcrumbs = character()) {
+  # short circuit on recursive environment traversal
+  if (env_name(x) %in% breadcrumbs)
+    return(NULL)
+
+  flat_map_srcrefs(
+    as.list(x, all.names = TRUE),
+    ns = env_ns_name(x),
+    breadcrumbs = c(breadcrumbs, env_name(x))
+  )
 }
 
 #' @exportS3Method
 #' @rdname srcrefs
-srcrefs.environment <- function(x, ...) {
-  if (isNamespace(x)) return(srcrefs.namespace(x))
+srcrefs.environment <- function(x, ..., breadcrumbs = character()) {
+  if (isNamespace(x))
+    return(srcrefs.namespace(x, ..., breadcrumbs = breadcrumbs))
+
+  # short circuit on recursive environment traversal
+  if (env_name(x) %in% breadcrumbs)
+    return(NULL)
+
   objs <- as.list(x, all.names = TRUE)
   objs <- Filter(function(i) !identical(i, x), objs)  # prevent direct recursion
-  flat_map_srcrefs(objs, ns = env_ns_name(x))
+  flat_map_srcrefs(
+    objs,
+    ns = env_ns_name(x),
+    breadcrumbs = c(breadcrumbs, env_name(x))
+  )
```

Comment:
This is the most important change.

Environment names are appended to `breadcrumbs` as we recurse. Before we recurse into the environment, we test to see if the environment name is part of our `breadcrumbs` vector. If it is, there's no need to recurse any further since we've already grabbed any non-environment srcrefs.

Reply:
Seems really simple and yet solves our problem. This is a great solution!